src/deploy/extensions/planner.ts

```diff
@@ -123,6 +126,64 @@ export async function have(projectId: string): Promise<DeploymentInstanceSpec[]>
   });
 }
 
+/**
+ * wantDynamic checks the users code for Extension SDKs usage and returns
+ * any extensions the user has defined that way.
+ * @param args
+ */
+export async function wantDynamic(args: {
+  projectId: string;
+  projectNumber: string;
+  aliases: string[];
+  projectDir: string;
+  extensions: Record<string, DynamicExtension>;
+  emulatorMode?: boolean;
+}): Promise<DeploymentInstanceSpec[]> {
+  const instanceSpecs: DeploymentInstanceSpec[] = [];
+  const errors: FirebaseError[] = [];
+  for (const [instanceId, ext] of Object.entries(args.extensions)) {
+    const autoPopulatedParams = await getFirebaseProjectParams(args.projectId, args.emulatorMode);
```

Comment:
A lot of this parameter handling logic is (nearly) duplicated here and in want(). Consider extracting into a separate function and calling it from both.

Reply:
I tried this, and other than the first 4 lines, it turned into a big mess of if(isDynamic) everywhere. Complicated by the different types of extension (one is Record<string, string>, one is Record<string, DynamicExtension>). So I don't think this actually would be better after all.